### ISSUE TITLE
estimateCellCounts2(...) fails with Error in UseMethod("filter_")

### ISSUE THREAD

**Issue by snowpong** [immunomethylomics/FlowSorted.Blood.EPIC#5]
I have an rgSet with EPIC data.

As the original minfi::estimateCellCounts(..) don't seem to work with FlowSorted.Blood.EPIC I figured I'd try your implementation of it. If fails with this error:

```
> estimateCellCounts2(rgSet, referencePlatform = "IlluminaHumanMethylationEPIC")
Error in UseMethod("filter_") : 
  no applicable method for 'filter_' applied to an object of class "c('tbl_SQLiteConnection', 'tbl_dbi', 'tbl_sql', 'tbl_lazy', 'tbl')"
```
Any ideas?
Note: The first time I ran this it asked for permission do write some local file cache which I allowed. Very strange.

If I run without specifying the referencePlatform it works fine (it converts to 450k and does it's job).

If I specify referenceSet it doesn't work:

```
> estimateCellCounts2(rgSet, referencePlatform = "IlluminaHumanMethylationEPIC", referenceset = "FlowSorted.Blood.EPIC")
Error in get(referenceset) : object 'FlowSorted.Blood.EPIC' not found
```

My guesses

-  some other package I have installed might be shadowing a function and your code is calling the wrong thing
- blocked network is stopping your function somehow (I'm running in a secure / restricted environment)

```
sessionInfo()
R version 3.6.3 (2020-02-29)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.7.1
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.7.1

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
 [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
[10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats4    parallel  stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] IlluminaHumanMethylation450kanno.ilmn12.hg19_0.6.0 
 [2] IlluminaHumanMethylation450kmanifest_0.4.0         
 [3] FlowSorted.Blood.450k_1.24.0                       
 [4] FlowSorted.Blood.EPIC_1.4.1                        
 [5] ExperimentHub_1.12.0                               
 [6] AnnotationHub_2.18.0                               
 [7] BiocFileCache_1.10.2                               
 [8] dbplyr_2.0.0                                       
 [9] nlme_3.1-150                                       
[10] quadprog_1.5-8                                     
[11] genefilter_1.68.0                                  
[12] IlluminaHumanMethylationEPICanno.ilm10b4.hg19_0.6.0
[13] IlluminaHumanMethylationEPICmanifest_0.3.0         
[14] minfi_1.32.0                                       
[15] bumphunter_1.28.0                                  
[16] locfit_1.5-9.4                                     
[17] iterators_1.0.13                                   
[18] foreach_1.5.1                                      
[19] Biostrings_2.54.0                                  
[20] XVector_0.26.0                                     
[21] SummarizedExperiment_1.16.1                        
[22] DelayedArray_0.12.3                                
[23] BiocParallel_1.20.1                                
[24] matrixStats_0.57.0                                 
[25] Biobase_2.46.0                                     
[26] GenomicRanges_1.38.0                               
[27] GenomeInfoDb_1.22.1                                
[28] IRanges_2.20.2                                     
[29] S4Vectors_0.24.4                                   
[30] BiocGenerics_0.32.0                                
[31] data.table_1.13.4                                  

loaded via a namespace (and not attached):
 [1] ellipsis_0.3.1                siggenes_1.60.0               mclust_5.4.7                 
 [4] base64_2.0                    rstudioapi_0.13               bit64_4.0.5                  
 [7] interactiveDisplayBase_1.24.0 AnnotationDbi_1.48.0          xml2_1.3.2                   
[10] codetools_0.2-18              splines_3.6.3                 scrime_1.3.5                 
[13] Rsamtools_2.2.3               annotate_1.64.0               shiny_1.5.0                  
[16] HDF5Array_1.14.4              BiocManager_1.30.10           readr_1.4.0                  
[19] compiler_3.6.3                httr_1.4.2                    assertthat_0.2.1             
[22] Matrix_1.2-18                 fastmap_1.0.1                 limma_3.42.2                 
[25] later_1.1.0.1                 htmltools_0.5.1               prettyunits_1.1.1            
[28] tools_3.6.3                   glue_1.4.2                    GenomeInfoDbData_1.2.2       
[31] dplyr_1.0.2                   rappdirs_0.3.1                doRNG_1.8.2                  
[34] Rcpp_1.0.5                    vctrs_0.3.5                   multtest_2.42.0              
[37] preprocessCore_1.48.0         rtracklayer_1.46.0            DelayedMatrixStats_1.8.0     
[40] stringr_1.4.0                 mime_0.9                      lifecycle_0.2.0              
[43] rngtools_1.5                  XML_3.99-0.3                  beanplot_1.2                 
[46] zlibbioc_1.32.0               MASS_7.3-53                   hms_0.5.3                    
[49] promises_1.1.1                rhdf5_2.30.1                  GEOquery_2.54.1              
[52] RColorBrewer_1.1-2            yaml_2.2.1                    curl_4.3                     
[55] memoise_1.1.0                 biomaRt_2.42.1                reshape_0.8.8                
[58] stringi_1.5.3                 RSQLite_2.2.1                 BiocVersion_3.10.1           
[61] GenomicFeatures_1.38.2        rlang_0.4.9                   pkgconfig_2.0.3              
[64] bitops_1.0-6                  nor1mix_1.3-0                 lattice_0.20-41              
[67] purrr_0.3.4                   Rhdf5lib_1.8.0                GenomicAlignments_1.22.1     
[70] bit_4.0.4                     tidyselect_1.1.0              plyr_1.8.6                   
[73] magrittr_2.0.1                R6_2.5.0                      generics_0.1.0               
[76] DBI_1.1.0                     pillar_1.4.7                  survival_3.2-7               
[79] RCurl_1.98-1.2                tibble_3.0.4                  crayon_1.3.4                 
[82] progress_1.2.2                grid_3.6.3                    blob_1.2.1                   
[85] digest_0.6.27                 xtable_1.8-4                  tidyr_1.1.2                  
[88] httpuv_1.5.5                  illuminaio_0.28.0             openssl_1.4.3                
[91] askpass_1.1        

```

**Comment by lucassalas** [comment]
Hi @snowpong ,

Let me clarify what happened here. We wrote this package with the intention of being used within the minfi environment. If you want to use minfi you can use it (see the code below). estimateCellCounts2 is an option if you want to run our IDOL library (which we highly recommend). We are offering this option as a temporary bypass, with some additional flexibility for other concepts that we developed in our manuscript. However, the original functions were developed by the minfi group as mentioned in the description and references, and we only adapted them to integrate our IDOL option.

Said so, have you run the example in ?estimateCellCounts2 ? is it working?

Try the code below:

```
# Step 1: Load the reference library
library(FlowSorted.Blood.EPIC)
library(ExperimentHub)
hub <- ExperimentHub()
query(hub, "FlowSorted.Blood.EPIC")
FlowSorted.Blood.EPIC <- hub[["EH1136"]]
FlowSorted.Blood.EPIC

# Step 2 separate the reference from the testing dataset if you want to run 
# examples for estimations for this function example **OR use your RGset**

RGsetTargets <- FlowSorted.Blood.EPIC[,
                                      FlowSorted.Blood.EPIC$CellType == "MIX"]

sampleNames(RGsetTargets) <- paste(RGsetTargets$CellType,
                                   seq_len(dim(RGsetTargets)[2]), sep = "_")
RGsetTargets

# **Step 3: use your favorite package for deconvolution.**
# Deconvolute a target data set consisting of EPIC DNA methylation 
# data profiled in blood, using your prefered method.

#If you want to use **minfi** default options, you can now use this code 
counts <- estimateCellCounts(RGsetTargets, processMethod = "auto", probeSelect = "auto",
                             cellTypes = c("CD8T","CD4T", "NK","Bcell","Mono","Neu"),
                             referencePlatform ="IlluminaHumanMethylationEPIC",
                             returnAll = FALSE)
round(counts, 2)
#There will be a warning because of an ExperimentHub issue, but it will work.

#HOWEVER
# You can use our **IDOL optimized DMR library for the EPIC array**.  This object
# contains a vector of length 450 consisting of the IDs of the IDOL optimized
# CpG probes.  These CpGs are used as the backbone for deconvolution and were
# selected because their methylation signature differs across the six normal 
# leukocyte subtypes.

data (IDOLOptimizedCpGs)
head (IDOLOptimizedCpGs)
# If you need to deconvolute a 450k legacy dataset use 
# IDOLOptimizedCpGs450klegacy instead

# We recommend using Noob processMethod = "preprocessNoob" in minfi for the 
# target and reference datasets. 
# Cell types included are "CD8T", "CD4T", "NK", "Bcell", "Mono", "Neu"

# To use the IDOL optimized list of CpGs (IDOLOptimizedCpGs) use 
# **estimateCellCounts2** an adaptation of the popular estimateCellCounts in 
# minfi. This function also allows including customized reference arrays. 

# Do not run with limited RAM the normalization step requires a big amount 
# of memory resources

if (memory.limit()>8000){
  countsEPIC<-estimateCellCounts2(RGsetTargets, compositeCellType = "Blood", 
                                  processMethod = "preprocessNoob",
                                  probeSelect = "IDOL", 
                                  cellTypes = c("CD8T", "CD4T", "NK", "Bcell", 
                                                "Mono", "Neu"), 
                                  referencePlatform = 
                                    "IlluminaHumanMethylationEPIC",
                                  referenceset = NULL,
                                  IDOLOptimizedCpGs =IDOLOptimizedCpGs, 
                                  returnAll = FALSE)
  
  head(countsEPIC$counts)
}

```
If you prefer to use the default minfi option for probe selection, for any reason, you can do so. We do not recommend that option by default as the estimates are less precise, but in some scenarios it is the only feasible alternative and you can use it.
I hope that helps you in taking an informed decision about what to do. Again, the IDOL option is only available through our function currently, that could be changed and adapted as required by other Bioconductor developers.

Also as a gentle reminder please reference in your manuscripts the original Houseman et al 2012, and Salas et al 2018 when using the CP/QP, and the FlowSorted.Blood.EPIC package. Please use the appropriate reference for the deconvolution method (IDOL Koestler et al. 2016, minfi pickCompProbes Jaffe et al 2014) and any other packages that you used. 

I will leave this issue open for one more day. Do all the steps in the example and you will be able to run the function.

**Comment by snowpong** [comment]
Thank you for the detailed reply.

I've looked a bit more into this.  The `Error in UseMethod("filter_")` comes from a [bug in AnnotationHub](https://github.com/Bioconductor/AnnotationHub/issues/21) which ExperimentHub depends on. It is fixed in Bioconductor 3.12, but that requires me moving to R version 4 - and I don't have time for that right now.
